packages/wrangler/src/experimental-assets.ts

```diff
@@ -1,8 +1,299 @@
+import { isJwtExpired } from "./pages/upload";
+import { APIError } from "./parse";
+import { urlSafe } from "./sites";
 import type { Config } from "./config";
 
+export type AssetManifest = { [path: string]: { hash: string; size: number } };
+
+type InitializeAssetsResponse = {
+	// string of file hashes per bucket
+	buckets: string[][];
+	jwt: string;
+};
+
+export type UploadPayloadFile = {
+	base64: boolean;
+	key: string;
+	value: string;
+	metadata: {
+		contentType: string;
+	};
+};
+
+type UploadResponse = {
+	jwt?: string;
+};
+
+// constants same as Pages for now
+const BULK_UPLOAD_CONCURRENCY = 3;
+const MAX_ASSET_COUNT = 20_000;
+const MAX_ASSET_SIZE = 25 * 1024 * 1024;
+const MAX_UPLOAD_ATTEMPTS = 5;
+const MAX_UPLOAD_GATEWAY_ERRORS = 5;
+
+export const syncExperimentalAssets = async (
+	accountId: string | undefined,
+	scriptName: string,
+	assetDirectory: string | undefined,
+	dryRun: boolean | undefined
+): Promise<string | undefined> => {
+	if (assetDirectory === undefined) {
+		return;
+	}
+	if (dryRun) {
+		logger.log("(Note: doing a dry run, not uploading or deleting anything.)");
+		return;
+	}
+	assert(accountId, "Missing accountId");
+
+	// 1. generate asset manifest
+	logger.info("🌀 Building list of assets...");
+	const manifest = await walk(assetDirectory, {});
+
+	// 2. fetch buckets w/ hashes
+	logger.info("🌀 Starting asset upload...");
+	const initializeAssetsResponse = await fetchResult<InitializeAssetsResponse>(
+		`/accounts/${accountId}/workers/scripts/${scriptName}/assets-upload-session`,
+		{
+			headers: { "Content-Type": "application/json" },
+			method: "POST",
+			body: JSON.stringify({ manifest: manifest }),
+		}
+	);
+
+	// if nothing to upload, return
+	if (initializeAssetsResponse.buckets.flat().length === 0) {
+		if (!initializeAssetsResponse.jwt) {
+			throw new FatalError(
+				"Could not find assets information to attach to deployment. Please try again.",
+				1
+			);
+		}
+		logger.info(`✨ Success! (No files to upload)`);
+		return initializeAssetsResponse.jwt;
+	}
+
+	// 3. fill buckets and upload assets
+	const numberFilesToUpload = initializeAssetsResponse.buckets.flat().length;
+	logger.info(`🌀 Uploading ${numberFilesToUpload} file(s)...`);
+
+	// Create the buckets outside of doUpload so we can retry without losing track of potential duplicate files
+	// But don't add the actual content until uploading so we don't run out of memory
+	const manifestLookup = Object.entries(manifest);
+	let assetLogCount = 0;
+	const bucketSkeletons = initializeAssetsResponse.buckets.map((bucket) => {
+		return bucket.map((fileHash) => {
+			const manifestEntry = manifestLookup.find(
+				(file) => file[1].hash === fileHash
+			);
+			if (manifestEntry === undefined) {
+				throw new FatalError(
+					`A file was requested that does not appear to exist.`,
+					1
+				);
+			}
+			// just logging file uploads at the moment...
+			// unsure how to log deletion vs unchanged file ignored/if we want to log this
```

Comment:
smth to figure out now or later? if for later, can we pls track this somewhere :pray:?

Reply:
tracking this in [this ticket](https://jira.cfdata.org/browse/WC-2464)